flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/BackendRestorerProcedure.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.flink.runtime.state.StateObject;
+import org.apache.flink.util.Disposable;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.ThrowingSupplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+
+/**
+ * This class implements the logic that creates (and potentially restores) a state backend. The restore logic
+ * considers multiple, prioritized options of snapshots to restore from, where all of the options should recreate
+ * the same state for the backend. When we fail to restore from the snapshot with the highest priority (typically
+ * the "fastest" to restore), we fallback to the next snapshot with the next highest priority. We also take care
+ * of cleaning up from failed restore attempts. We only reattempt when the problem occurs during the restore call
+ * and will only stop after all snapshot alternatives are exhausted and all failed.
+ *
+ * @param <T> type of the restored backend.
+ * @param <S> type of the supplied snapshots from which the backend restores.
+ */
+public class BackendRestorerProcedure<
+	T extends Closeable & Disposable & Snapshotable<?, StateObjectCollection<S>>,
+	S extends StateObject> {
+
+	/** Logger for this class. */
+	private static final Logger LOG = LoggerFactory.getLogger(BackendRestorerProcedure.class);
+
+	/** Factory for new, fresh backends without state. */
+	private final ThrowingSupplier<T, Exception> instanceSupplier;
+
+	/** This registry is used so that recovery can participate in the task lifecycle, i.e. can be canceled. */
+	private final CloseableRegistry backendCloseableRegistry;
+
+	/**
+	 * Creates a new backend restorer using the given backend supplier and the closeable registry.
+	 *
+	 * @param instanceSupplier factory function for new, empty backend instances.
+	 * @param backendCloseableRegistry registry to allow participation in task lifecycle, e.g. react to cancel.
+	 */
+	public BackendRestorerProcedure(
+		@Nonnull ThrowingSupplier<T, Exception> instanceSupplier,
+		@Nonnull CloseableRegistry backendCloseableRegistry) {
+
+		this.instanceSupplier = Preconditions.checkNotNull(instanceSupplier);
+		this.backendCloseableRegistry = Preconditions.checkNotNull(backendCloseableRegistry);
+	}
+
+	/**
+	 * Creates a new state backend and restores it from the provided set of state snapshot alternatives.
+	 *
+	 * @param restoreOptions iterator over a prioritized set of state snapshot alternatives for recovery.
+	 * @return the created (and restored) state backend.
+	 * @throws Exception if the backend could not be created or restored.
+	 */
+	public @Nonnull T createAndRestore(@Nonnull Iterator<StateObjectCollection<S>> restoreOptions) throws Exception {
+
+		StateObjectCollection<S> restoreState = null;
+
+		boolean retry;
+
+		while (true) {
```

Comment:
I think we could get rid of the endless loop here by simply using the iterator.
```
while (restoreOptions.hasNext()) {
  final StateObjectCollection<S> restoreState = restoreOptions.next();

  try {
     ...
  } catch (Exception e) {
    ...
  }
}

throw new FlinkException("Could not restore from any of the provided restore options");
```

Reply:
It is not quiet that easy, because the restore method must always be called at least once and also the log message might differ between iterations (the "will retry" part). But I agree it could be simplified.